bitcoinetl/jobs/export_blocks_job.py

```diff
@@ -0,0 +1,119 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+import json
+
+from blockchainetl.executors.batch_work_executor import BatchWorkExecutor
+from blockchainetl.jobs.base_job import BaseJob
+from blockchainetl.utils import rpc_response_batch_to_results, validate_range
+from bitcoinetl.mappers.block_mapper import BtcBlockMapper
+from bitcoinetl.json_rpc_requests import generate_get_block_by_hash_json_rpc, generate_get_block_hash_by_number_json_rpc, generate_get_transaction_by_id_json_rpc
+from bitcoinetl.mappers.transaction_mapper import BtcTransactionMapper
+
+
+# Exports blocks and transactions
+class ExportBlocksJob(BaseJob):
+    def __init__(
+            self,
+            start_block,
+            end_block,
+            batch_size,
+            batch_rpc_provider,
+            max_workers,
+            item_exporter,
+            export_blocks=True,
+            export_transactions=True):
+        validate_range(start_block, end_block)
+        self.start_block = start_block
+        self.end_block = end_block
+
+        self.rpc_provider = batch_rpc_provider
+
+        self.batch_work_executor = BatchWorkExecutor(batch_size, max_workers)
+        self.item_exporter = item_exporter
+
+        self.export_blocks = export_blocks
+        self.export_transactions = export_transactions
+        if not self.export_blocks and not self.export_transactions:
+            raise ValueError('At least one of export_blocks or export_transactions must be True')
+
+        self.block_mapper = BtcBlockMapper()
+        self.transaction_mapper = BtcTransactionMapper()
+
+    def _start(self):
+        self.item_exporter.open()
+
+    def _export(self):
+        self.batch_work_executor.execute(
+            range(self.start_block, self.end_block + 1),
+            self._export_batch,
+            total_items=self.end_block - self.start_block + 1
+        )
+
+    def add_input_values(self, block):
+        for txIndex, tx in enumerate(block['tx']):
+            txids = list(map(lambda vin: vin["txid"] if "txid" in vin else "", tx['vin']))
+
+            transaction_detail_rpc = list(generate_get_transaction_by_id_json_rpc(txids))
+            transaction_detail_response = self.rpc_provider.make_request(transaction_detail_rpc)
```

Comment:
I'm not sure JSON RPC batch responses are guaranteed to be returned in the same order as requests in the batch (seems below you rely on it). From the spec examples it looks like the order can be arbitrary and the "id" field needs to be used to match up requests and responses https://www.jsonrpc.org/specification#batch.

Reply:
I'm not sure just adding ids to the request will guarantee the responses will come back in the same order. In the JSON RPC example here https://www.jsonrpc.org/specification#batch the requests come with the ids in a certain order, but the responses are returned in an arbitrary order:

```--> [
        {"jsonrpc": "2.0", "method": "sum", "params": [1,2,4], "id": "1"},
        {"jsonrpc": "2.0", "method": "notify_hello", "params": [7]},
        {"jsonrpc": "2.0", "method": "subtract", "params": [42,23], "id": "2"},
        {"foo": "boo"},
        {"jsonrpc": "2.0", "method": "foo.get", "params": {"name": "myself"}, "id": "5"},
        {"jsonrpc": "2.0", "method": "get_data", "id": "9"} 
    ]
<-- [
        {"jsonrpc": "2.0", "result": 7, "id": "1"},
        {"jsonrpc": "2.0", "result": 19, "id": "2"},
        {"jsonrpc": "2.0", "error": {"code": -32600, "message": "Invalid Request"}, "id": null},
        {"jsonrpc": "2.0", "error": {"code": -32601, "message": "Method not found"}, "id": "5"},
        {"jsonrpc": "2.0", "result": ["hello", 5], "id": "9"}
    ]
```

In `pythonbitcoinrpc` they just put the incremented ids to requests and then never use them (e.g. to order the responses). Even worse there is no way for us to provide custom ids and handle it ourselves.

A simple way to handle it is to sort the responses by id. I will also submit an issue to pythonbitcoinrpc.

Reply:
@medvedev1088 Yeah true. I would just include include the `authproxy.py` file directly in the codebase and make the change to sort the response by id. Would remove the pythonbitcoinrpc dependency.